### ISSUE TITLE
Optional mapper settings are 

### ISSUE THREAD

**Issue by johnnovak** [dosbox-staging/dosbox-staging#2771]
# Problem

Some mapper events are optional in the sense that they are dependent on the configured hardware. For example, the below mapper events are only available for the `hercules` and `cga-mono` machine types, respectively:

- "Herc Pal"
- "Mono CGA Pal"
- "Mono CGA Bright"

In any other modes, these events are simply not created, therefore the mapper doesn't load them, which means on subsequent saving of the mapper file they are dropped.

# How to reproduce

1. Start DOSBox in `machine = hercules` mode.
1. Open the mapper and assign the "Herc Pal" even to some key (let's say F11).
1. Save the mapper settings.
1. Open the mapper file and confirm the following entry is preset: `hand_hercpal "key 68"`  (key 68 = F11)

---

1. Now start DOSBox with any other machine type.
1. Open the mapper — note that "Herc Pal" is not present, which is good and expected as it's for Hercules only.
1. Save the mapper settings without making any changes to it.
1. Open the mapper file again. Notice that the `hand_hercpal` entry is gone.

# Solution

The whole mapper model is flawed. The problem happens because optional mappings are only created optionally at runtime. No mapping created means it's invisible to the mapper, so those settings are dropped.

A better model would be that all possible events are registered/created at all time (once per startup), then they could be dynamically enabled/disabled as needed. So the Hercules and Mono CGA related events would be always created internally in the mapper, no matter what, and all possible events, naturally. Then they would be only enabled via an "enabled" flag or something when entering a Hercules or CGA mono mode, respectively.

Not sure if messing around with the current mapper is worth it, though. But something to think about for the future mapper. Nevertheless, the current behaviour is super confusing for our users presently. Basically all such conditional settings would be randomly lost whenever you save the mapping file (yeah, it's not random at all, but only we know that... for most users, this will appear as random flaky behaviour).

**Comment by kcgen** [comment]
Totally agree 👍; I suppose the mapper's hotkey action/function registration could include an enable/disable callback that can re-evaluate if the function should be available or not, each time the mapper is re-opened.

I also dunno if it's worth it to heap that onto the current mapper; but it's great to have this need in mind during the re-write.

One of the least frustrating ways to manage mappings in its current form is to save game-specific mappings to a local mapper file, along side the game's dosbox.conf.

This is how exo sets up normal keyboard arrow-key and action buttons mappings for some games that were joystick-only (about 40 of these games, each with their own mapper file).

Only mentioning this because this per-config use-case kind of side steps this issue (which is more a problem using one mapper file for various machine types).

**Comment by johnnovak** [comment]
> I also dunno if it's worth it to heap that onto the current mapper; but it's great to have this need in mind during the re-write.

Nah, we shouldn't spend any effort on the old mapper unless it's either trivial stuff, or something is horribly broken and there's no workaround. I guess I just wanted to document this as a "known issue".

> Only mentioning this because this per-config use-case kind of side steps this issue (which is more a problem using one mapper file for various machine types).

Yeah, that's a fine workaround in practice, I think. It just constantly annoys me when switching between testing Hercules and CGA mono modes.

**Comment by kcgen** [comment]
> Yeah, that's a fine workaround in practice, I think. It just constantly annoys me when switching between testing Hercules and CGA mono modes.

Oof; yeah.

Maybe a short-term work around is to launch with a `-conf herc.conf` or `-conf cga.conf` both of which tweak the `machine=` and `mapperfile=` values.

**Comment by johnnovak** [comment]
> > Yeah, that's a fine workaround in practice, I think. It just constantly annoys me when switching between testing Hercules and CGA mono modes.
> 
> Oof; yeah.
> 
> Maybe a short-term work around is to launch with a `-conf herc.conf` or `-conf cga.conf` both of which tweak the `machine=` and `mapperfile=` values.

Yeah that would work. I just got really efficient at using the mapper 😆
